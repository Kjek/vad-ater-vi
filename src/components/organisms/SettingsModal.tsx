import InputButton from '@component/admin/Button';
import IconButton from '@component/atoms/IconButton';
import Text from '@component/atoms/Text';
import { api } from '@util/api';
import { toastError, toastSuccessful } from '@util/toast-utils';
import { useRef } from 'react';
import type { Id } from 'react-toastify';
import { toast } from 'react-toastify';
import { useToggle } from 'usehooks-ts';
import SettingsItem from '../molecules/SettingsItem';

interface SettingsModalProps {
  restaurant: {
    id: string;
    name: string;
    homeUrl: string;
    lunchUrl: string;
    regex: string | null;
  };
}

const SettingsModal = (props: SettingsModalProps) => {
  const { restaurant } = props;
  const [open, toggleOpen] = useToggle();
  const toastId = useRef<Id>(0);
  const restaurantName = useRef<string>(restaurant.name);
  const homeUrl = useRef<string>(restaurant.homeUrl);
  const lunchUrl = useRef<string>(restaurant.lunchUrl);
  const regex = useRef<string>(restaurant.regex || '');
  const { mutate: updateSettings } =
    api.admin.updateRestaurantSetting.useMutation({
      onSuccess() {
        toastSuccessful(toastId.current);
      },
      onError(error) {
        toastError(toastId.current, error);
      },
      onMutate() {
        toastId.current = toast.loading('Updating in progress...');
      },
    });

  const handleUpdate = () => {
    updateSettings({
      id: restaurant.id,
      name: restaurantName.current,
      homeUrl: homeUrl.current,
      lunchUrl: lunchUrl.current,
      regex: regex.current,
    });
  };

  return (
    <>
      <Text className='flex-1 cursor-pointer' onClick={toggleOpen}>
        {restaurant.name}
      </Text>
      {open ? (
        <>
          <div className='absolute left-0 top-0 h-screen w-screen bg-black opacity-30' />
          <div className='dark:bg-gray-custom fixed left-[50%] top-[50%] flex max-h-[85vh] w-[90vw] max-w-[450px] translate-x-[-50%] translate-y-[-50%] flex-col gap-4 rounded-[6px] bg-white p-6'>
            <Text variant='h3'>Edit restaurant</Text>
            <Text variant='h5'>
              {
                "Make changes to the restaurant here. Click save when you're done."
              }
            </Text>
            <div>
              <SettingsItem
                title='Name'
                inputDefaultValue={restaurantName.current}
                inputPlaceholder='Name'
                onChange={(event) =>
                  (restaurantName.current = event.target.value)
                }
              />
              <SettingsItem
                title='Home URL'
                inputDefaultValue={homeUrl.current}
                inputPlaceholder='https://example.com'
                onChange={(event) => (homeUrl.current = event.target.value)}
              />
              <SettingsItem
                title='Lunch URL'
                inputDefaultValue={lunchUrl.current}
                inputPlaceholder='https://example.com/lunch'
                onChange={(event) => (lunchUrl.current = event.target.value)}
              />
              <SettingsItem
                title='Regex'
                inputDefaultValue={regex.current}
                inputPlaceholder='/^([a-z])$/gm'
                onChange={(event) => (regex.current = event.target.value)}
              />
              <div className='mt-6 flex justify-end'>
                <InputButton
                  value='Save changes'
                  aria-label='Save'
                  onClick={handleUpdate}
                />
              </div>
            </div>
            <IconButton
              variant='cancel'
              className='absolute right-2 top-2 p-2'
              onClick={toggleOpen}
            />
          </div>
        </>
      ) : null}
    </>
  );
};

export default SettingsModal;
