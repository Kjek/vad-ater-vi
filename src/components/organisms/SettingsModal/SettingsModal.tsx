import InputButton from '@component/atoms/Button/Button';
import IconButton from '@component/atoms/IconButton/IconButton';
import Text from '@component/atoms/Text/Text';
import SettingsItem from '@component/molecules/SettingsItem/SettingsItem';
import { api } from '@util/api';
import { toastError, toastSuccessful } from '@util/toast-utils';
import { useRef } from 'react';
import type { Id } from 'react-toastify';
import { toast } from 'react-toastify';
import { useToggle } from 'usehooks-ts';

interface SettingsModalProps {
  restaurant: {
    id: string;
    name: string;
    homeUrl: string;
    lunchUrl: string;
    lunchRegex: string | null;
    weeklyRegex: string | null;
    restaurantId: string;
  };
}

const SettingsModal = (props: SettingsModalProps) => {
  const { restaurant } = props;
  const [open, toggleOpen] = useToggle();
  const toastId = useRef<Id>(0);
  const restaurantName = useRef<string>(restaurant.name);
  const homeUrl = useRef<string>(restaurant.homeUrl);
  const lunchUrl = useRef<string>(restaurant.lunchUrl);
  const lunchRegex = useRef<string>(restaurant.lunchRegex || '');
  const weeklyRegex = useRef<string>(restaurant.weeklyRegex || '');
  const { mutate: updateSettings } =
    api.admin.updateRestaurantSetting.useMutation({
      onSuccess() {
        toastSuccessful(toastId.current);
      },
      onError(error) {
        toastError(toastId.current, error);
      },
      onMutate() {
        toastId.current = toast.loading('Updating in progress...');
      },
    });

  const { mutate: deleteRestaurant } = api.admin.deleteRestaurant.useMutation({
    onSuccess() {
      toastSuccessful(toastId.current);
    },
    onError(error) {
      toastError(toastId.current, error);
    },
    onMutate() {
      toastId.current = toast.loading('Updating in progress...');
    },
  });

  const handleUpdate = () => {
    updateSettings({
      id: restaurant.id,
      name: restaurantName.current,
      homeUrl: homeUrl.current,
      lunchUrl: lunchUrl.current,
      lunchRegex: lunchRegex.current,
    });
  };

  const handleDelete = () => {
    deleteRestaurant({ id: restaurant.restaurantId });
    toggleOpen();
  };

  return (
    <>
      <Text className='center flex-1 cursor-pointer' onClick={toggleOpen}>
        {restaurant.name}
      </Text>
      {open ? (
        <>
          <div className='absolute left-0 top-0 h-screen w-full bg-black opacity-30' />
          <div className='dark:bg-gray-custom fixed left-[50%] top-[50%] flex max-h-[85vh] w-[90vw] max-w-[450px] translate-x-[-50%] translate-y-[-50%] flex-col gap-4 rounded-[6px] bg-white p-6'>
            <Text variant='h3'>Edit restaurant</Text>
            <Text variant='h5'>
              {
                "Make changes to the restaurant here. Click save when you're done."
              }
            </Text>
            <div>
              <SettingsItem
                title='Name'
                inputDefaultValue={restaurantName.current}
                inputPlaceholder='Name'
                onChange={(event) =>
                  (restaurantName.current = event.target.value)
                }
              />
              <SettingsItem
                title='Home URL'
                inputDefaultValue={homeUrl.current}
                inputPlaceholder='https://example.com'
                onChange={(event) => (homeUrl.current = event.target.value)}
              />
              <SettingsItem
                title='Lunch URL'
                inputDefaultValue={lunchUrl.current}
                inputPlaceholder='https://example.com/lunch'
                onChange={(event) => (lunchUrl.current = event.target.value)}
              />
              <SettingsItem
                title='Lunch Regex'
                inputDefaultValue={lunchRegex.current}
                inputPlaceholder='/^([a-z])$/gm'
                onChange={(event) => (lunchRegex.current = event.target.value)}
              />
              <SettingsItem
                title='Weekly Regex'
                inputDefaultValue={weeklyRegex.current}
                inputPlaceholder='/^([a-z])$/gm'
                onChange={(event) => (weeklyRegex.current = event.target.value)}
              />
              <div className='mt-6 flex gap-4'>
                <div className='flex-1'>
                  <InputButton
                    value='Delete'
                    aria-label='Delete'
                    onClick={handleDelete}
                    className='text-red-500 dark:text-red-700'
                  />
                </div>
                <InputButton
                  value='Save changes'
                  aria-label='Save'
                  onClick={handleUpdate}
                />
              </div>
            </div>
            <IconButton
              variant='cancel'
              className='absolute right-2 top-2 p-2'
              onClick={toggleOpen}
            />
          </div>
        </>
      ) : null}
    </>
  );
};

export default SettingsModal;
