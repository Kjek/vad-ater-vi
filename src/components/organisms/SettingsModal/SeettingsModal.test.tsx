import { render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import SettingsModal from './SettingsModal';

String.prototype.toDotNotation = vi.fn();

describe(SettingsModal, () => {
  it('renders correctly', () => {
    const { container } = render(
      <SettingsModal
        deleteRestaurant={vi.fn()}
        restaurant={{
          id: '0',
          name: 'Example',
          homeUrl: 'www.example.com',
          lunchUrl: 'www.example.com/lunch',
          lunchRegex: null,
          weeklyRegex: null,
          enabled: true,
          restaurantId: '0',
        }}
        toggleOpen={vi.fn()}
        updateSettings={vi.fn()}
      />
    );
    expect(container).toBeInTheDocument();
    expect(container).toBeVisible();
  });

  it('matches snapshot', () => {
    const defaultRender = render(
      <SettingsModal
        deleteRestaurant={vi.fn()}
        restaurant={{
          id: '0',
          name: 'Example',
          homeUrl: 'www.example.com',
          lunchUrl: 'www.example.com/lunch',
          lunchRegex: null,
          weeklyRegex: null,
          enabled: true,
          restaurantId: '0',
        }}
        toggleOpen={vi.fn()}
        updateSettings={vi.fn()}
      />
    );
    expect(defaultRender).toMatchSnapshot();
  });
});
