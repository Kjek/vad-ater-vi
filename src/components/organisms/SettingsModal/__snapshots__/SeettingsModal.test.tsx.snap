// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsModal > matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fixed left-0 top-0 m-0 h-full w-full bg-black opacity-30"
      />
      <div
        class="dark:bg-gray-custom fixed left-[50%] top-[50%] flex max-h-[85vh] w-[90vw] max-w-[450px] translate-x-[-50%] translate-y-[-50%] flex-col gap-4 rounded-[6px] bg-white p-6 pt-10 flex-grow self-center"
      >
        <h3
          class="text-2xl font-bold text-gray-800 dark:text-gray-300"
        >
          Edit resstaurant
        </h3>
        <h5
          class="text-gray-500 dark:text-gray-300"
        >
          Make changes to the restaurant here. Click save when you're done.
        </h5>
        <div>
          <div
            class="mb-4 flex items-center gap-4"
          >
            <h4
              class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
            >
              Name
            </h4>
            <input
              class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
              placeholder="Name"
              type="text"
              value="Example"
            />
          </div>
          <div
            class="mb-4 flex items-center gap-4"
          >
            <h4
              class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
            >
              Home URL
            </h4>
            <input
              class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
              placeholder="https://example.com"
              type="text"
              value="www.example.com"
            />
          </div>
          <div
            class="mb-4 flex items-center gap-4"
          >
            <h4
              class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
            >
              Lunch URL
            </h4>
            <input
              class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
              placeholder="https://example.com/lunch"
              type="text"
              value="www.example.com/lunch"
            />
          </div>
          <div
            class="mb-4 flex items-center gap-4"
          >
            <h4
              class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
            >
              Lunch Regex
            </h4>
            <input
              class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
              placeholder="/^([a-z])$/gm"
              type="text"
              value=""
            />
          </div>
          <div
            class="mb-4 flex items-center gap-4"
          >
            <h4
              class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
            >
              Weekly Regex
            </h4>
            <input
              class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
              placeholder="/^([a-z])$/gm"
              type="text"
              value=""
            />
          </div>
          <div
            class="mt-6 flex gap-4"
          >
            <div
              class="flex-1"
            >
              <input
                aria-label="Delete"
                class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium hover:bg-gray-200 dark:border-gray-600 dark:hover:bg-gray-700 dark:hover:text-white text-red-500 dark:text-red-700"
                type="button"
                value="Delete"
              />
            </div>
            <input
              aria-label="Save"
              class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white"
              type="button"
              value="Save changes"
            />
          </div>
        </div>
        <button
          class="rounded-full hover:bg-gray-200 dark:hover:bg-gray-700 absolute right-2 top-2 p-2"
          type="button"
        >
          <svg
            class="text-red-500 dark:text-red-700"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.8536 2.85355C13.0488 2.65829 13.0488 2.34171 12.8536 2.14645C12.6583 1.95118 12.3417 1.95118 12.1464 2.14645L7.5 6.79289L2.85355 2.14645C2.65829 1.95118 2.34171 1.95118 2.14645 2.14645C1.95118 2.34171 1.95118 2.65829 2.14645 2.85355L6.79289 7.5L2.14645 12.1464C1.95118 12.3417 1.95118 12.6583 2.14645 12.8536C2.34171 13.0488 2.65829 13.0488 2.85355 12.8536L7.5 8.20711L12.1464 12.8536C12.3417 13.0488 12.6583 13.0488 12.8536 12.8536C13.0488 12.6583 13.0488 12.3417 12.8536 12.1464L8.20711 7.5L12.8536 2.85355Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="fixed left-0 top-0 m-0 h-full w-full bg-black opacity-30"
    />
    <div
      class="dark:bg-gray-custom fixed left-[50%] top-[50%] flex max-h-[85vh] w-[90vw] max-w-[450px] translate-x-[-50%] translate-y-[-50%] flex-col gap-4 rounded-[6px] bg-white p-6 pt-10 flex-grow self-center"
    >
      <h3
        class="text-2xl font-bold text-gray-800 dark:text-gray-300"
      >
        Edit resstaurant
      </h3>
      <h5
        class="text-gray-500 dark:text-gray-300"
      >
        Make changes to the restaurant here. Click save when you're done.
      </h5>
      <div>
        <div
          class="mb-4 flex items-center gap-4"
        >
          <h4
            class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
          >
            Name
          </h4>
          <input
            class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
            placeholder="Name"
            type="text"
            value="Example"
          />
        </div>
        <div
          class="mb-4 flex items-center gap-4"
        >
          <h4
            class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
          >
            Home URL
          </h4>
          <input
            class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
            placeholder="https://example.com"
            type="text"
            value="www.example.com"
          />
        </div>
        <div
          class="mb-4 flex items-center gap-4"
        >
          <h4
            class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
          >
            Lunch URL
          </h4>
          <input
            class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
            placeholder="https://example.com/lunch"
            type="text"
            value="www.example.com/lunch"
          />
        </div>
        <div
          class="mb-4 flex items-center gap-4"
        >
          <h4
            class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
          >
            Lunch Regex
          </h4>
          <input
            class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
            placeholder="/^([a-z])$/gm"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-4 flex items-center gap-4"
        >
          <h4
            class="font-bold text-gray-800 dark:text-gray-300 w-2/5 text-sm"
          >
            Weekly Regex
          </h4>
          <input
            class="w-full self-center rounded-lg border border-gray-300 bg-gray-50 p-2.5 text-sm text-gray-800 outline-none focus:border-gray-300 focus:ring-gray-300 dark:border-gray-600 dark:bg-gray-800 dark:text-gray-300 dark:placeholder-gray-400 dark:focus:border-gray-600 dark:focus:ring-gray-600"
            placeholder="/^([a-z])$/gm"
            type="text"
            value=""
          />
        </div>
        <div
          class="mt-6 flex gap-4"
        >
          <div
            class="flex-1"
          >
            <input
              aria-label="Delete"
              class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium hover:bg-gray-200 dark:border-gray-600 dark:hover:bg-gray-700 dark:hover:text-white text-red-500 dark:text-red-700"
              type="button"
              value="Delete"
            />
          </div>
          <input
            aria-label="Save"
            class="cursor-pointer rounded-lg border border-gray-300 px-5 py-2.5 text-sm font-medium text-gray-800 hover:bg-gray-200 dark:border-gray-600 dark:text-gray-300 dark:hover:bg-gray-700 dark:hover:text-white"
            type="button"
            value="Save changes"
          />
        </div>
      </div>
      <button
        class="rounded-full hover:bg-gray-200 dark:hover:bg-gray-700 absolute right-2 top-2 p-2"
        type="button"
      >
        <svg
          class="text-red-500 dark:text-red-700"
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.8536 2.85355C13.0488 2.65829 13.0488 2.34171 12.8536 2.14645C12.6583 1.95118 12.3417 1.95118 12.1464 2.14645L7.5 6.79289L2.85355 2.14645C2.65829 1.95118 2.34171 1.95118 2.14645 2.14645C1.95118 2.34171 1.95118 2.65829 2.14645 2.85355L6.79289 7.5L2.14645 12.1464C1.95118 12.3417 1.95118 12.6583 2.14645 12.8536C2.34171 13.0488 2.65829 13.0488 2.85355 12.8536L7.5 8.20711L12.1464 12.8536C12.3417 13.0488 12.6583 13.0488 12.8536 12.8536C13.0488 12.6583 13.0488 12.3417 12.8536 12.1464L8.20711 7.5L12.8536 2.85355Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
